std/math.d

```diff
@@ -7318,3 +7319,286 @@ unittest
         }
     }
 }
+
+private enum PowType
+{
+    floor,
+    ceil
+}
+
+pragma(inline, true)
+private T powIntegralImpl(PowType type, T)(T val)
+{
+    import core.bitop : bsr;
+
+    Unqual!T x = val;
+
+    static if (isSigned!T)
+    if (x < 0)
+        x = abs(x);
+
+    if (x == 0)
+        return x;
+
+    static if (type == PowType.ceil)
+        x = cast(Unqual!T) (Unqual!T(1) << bsr(x) + 1);
+    else
+        x = cast(Unqual!T) (Unqual!T(1) << bsr(x));
+
+    static if (isSigned!T)
+    if (val < 0)
+        return -x;
+
+    return x;
+}
+
+pragma(inline, true)
+private T powFloatingPointImpl(PowType type, T)(T x)
+{
+    if (!x.isFinite)
+        return x;
+
+    if (!x)
+        return x;
+
+    int exp;
+    auto y = frexp(x, exp);
+
+    static if (type == PowType.ceil)
+        y = ldexp(cast(T) 0.5, exp + 1);
+    else
+        y = ldexp(cast(T) 0.5, exp);
+
+    if (!y.isFinite)
+        return cast(T) 0.0;
+
+    y = copysign(y, x);
+
+    return y;
+}
+
+/**
+ * Gives the next power of two after $(D val). $(T) can be any built-in
+ * numerical type.
+ *
+ * Params:
+ *     val = any number
+ *
+ * Returns:
+ *     the next power of two after $(D val)
+ */
+T nextPow2(T)(const T val) if (isIntegral!T)
+{
+    return powIntegralImpl!(PowType.ceil)(val);
+}
+
+/// ditto
+T nextPow2(T)(const T val) if (isFloatingPoint!T)
+{
+    return powFloatingPointImpl!(PowType.ceil)(val);
+}
+
+///
+@safe @nogc pure nothrow unittest
+{
+    assert(nextPow2(2) == 4);
+    assert(nextPow2(10) == 16);
+    assert(nextPow2(4000) == 4096);
+
+    assert(nextPow2(-2) == -4);
+    assert(nextPow2(-10) == -16);
+
+    assert(nextPow2(uint.max) == 1);
```

Comment:
Thanks for the explanation. Unfortunately the current implementation depends on processor behaviour.
I think it would be more consistent and reliable to return 0 for the "out of bounds" cases, similar to what happens for floating point values. This should just add one more comparison.